# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [1]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import accuracy_score
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor,AdaBoostRegressor, GradientBoostingRegressor

from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

In [3]:
spaceship.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


In [4]:
spaceship.shape

(8693, 14)

In [5]:
spaceship.isnull().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

In [6]:
spaceship_cleaned = spaceship.dropna()

In [7]:
spaceship_cleaned.isnull().sum()

PassengerId     0
HomePlanet      0
CryoSleep       0
Cabin           0
Destination     0
Age             0
VIP             0
RoomService     0
FoodCourt       0
ShoppingMall    0
Spa             0
VRDeck          0
Name            0
Transported     0
dtype: int64

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [8]:
# Transformar la columna "Cabin"
spaceship_cleaned['Cabin'] = spaceship_cleaned['Cabin'].apply(lambda x: x[0] if pd.notna(x) else 'Unknown')

# Mostrar el DataFrame transformado
print("\nDataFrame transformado:")
print(spaceship_cleaned)


DataFrame transformado:
     PassengerId HomePlanet CryoSleep Cabin    Destination   Age    VIP  \
0        0001_01     Europa     False     B    TRAPPIST-1e  39.0  False   
1        0002_01      Earth     False     F    TRAPPIST-1e  24.0  False   
2        0003_01     Europa     False     A    TRAPPIST-1e  58.0   True   
3        0003_02     Europa     False     A    TRAPPIST-1e  33.0  False   
4        0004_01      Earth     False     F    TRAPPIST-1e  16.0  False   
...          ...        ...       ...   ...            ...   ...    ...   
8688     9276_01     Europa     False     A    55 Cancri e  41.0   True   
8689     9278_01      Earth      True     G  PSO J318.5-22  18.0  False   
8690     9279_01      Earth     False     G    TRAPPIST-1e  26.0  False   
8691     9280_01     Europa     False     E    55 Cancri e  32.0  False   
8692     9280_02     Europa     False     E    TRAPPIST-1e  44.0  False   

      RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               N

C:\Users\evacv\AppData\Local\Temp\ipykernel_2700\1604076933.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  spaceship_cleaned['Cabin'] = spaceship_cleaned['Cabin'].apply(lambda x: x[0] if pd.notna(x) else 'Unknown')


In [9]:
#your code here
spaceship_cleaned = spaceship_cleaned.drop(columns=['PassengerId', 'Name'])

In [10]:
spaceship_cleaned = pd.get_dummies(spaceship_cleaned, columns=spaceship_cleaned.select_dtypes(include=['object']).columns)
spaceship_cleaned

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0     39.0          0.0        0.0           0.0     0.0     0.0        False   
1     24.0        109.0        9.0          25.0   549.0    44.0         True   
2     58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3     33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4     16.0        303.0       70.0         151.0   565.0     2.0         True   
...    ...          ...        ...           ...     ...     ...          ...   
8688  41.0          0.0     6819.0           0.0  1643.0    74.0        False   
8689  18.0          0.0        0.0           0.0     0.0     0.0        False   
8690  26.0          0.0        0.0        1872.0     1.0     0.0         True   
8691  32.0          0.0     1049.0           0.0   353.0  3235.0        False   
8692  44.0        126.0     4688.0           0.0     0.0    12.0         True   

      HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  ...  Cabin_D  \
0                False               True            False  ...    False   
1                 True              False            False  ...    False   
2                False               True            False  ...    False   
3                False               True            False  ...    False   
4                 True              False            False  ...    False   
...                ...                ...              ...  ...      ...   
8688             False               True            False  ...    False   
8689              True              False            False  ...    False   
8690              True              False            False  ...    False   
8691             False               True            False  ...    False   
8692             False               True            False  ...    False   

      Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0       False    False    False    False                    False   
1       False     True    False    False                    False   
2       False    False    False    False                    False   
3       False    False    False    False                    False   
4       False     True    False    False                    False   
...       ...      ...      ...      ...                      ...   
8688    False    False    False    False                     True   
8689    False    False     True    False                    False   
8690    False    False     True    False                    False   
8691     True    False    False    False                     True   
8692     True    False    False    False                    False   

      Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                         False                     True       True     False  
1                         False                     True       True     False  
2                         False                     True      False      True  
3                         False                     True       True     False  
4                         False                     True       True     False  
...                         ...                      ...        ...       ...  
8688                      False                    False      False      True  
8689                       True                    False       True     False  
8690                      False                     True       True     False  
8691                      False                    False       True     False  
8692                      False                     True       True     False  

[6606 rows x 25 columns]

In [11]:
numerical_spaceship = spaceship_cleaned.select_dtypes(include=['float64','bool'])
numerical_spaceship

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0     39.0          0.0        0.0           0.0     0.0     0.0        False   
1     24.0        109.0        9.0          25.0   549.0    44.0         True   
2     58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3     33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4     16.0        303.0       70.0         151.0   565.0     2.0         True   
...    ...          ...        ...           ...     ...     ...          ...   
8688  41.0          0.0     6819.0           0.0  1643.0    74.0        False   
8689  18.0          0.0        0.0           0.0     0.0     0.0        False   
8690  26.0          0.0        0.0        1872.0     1.0     0.0         True   
8691  32.0          0.0     1049.0           0.0   353.0  3235.0        False   
8692  44.0        126.0     4688.0           0.0     0.0    12.0         True   

      HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  ...  Cabin_D  \
0                False               True            False  ...    False   
1                 True              False            False  ...    False   
2                False               True            False  ...    False   
3                False               True            False  ...    False   
4                 True              False            False  ...    False   
...                ...                ...              ...  ...      ...   
8688             False               True            False  ...    False   
8689              True              False            False  ...    False   
8690              True              False            False  ...    False   
8691             False               True            False  ...    False   
8692             False               True            False  ...    False   

      Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0       False    False    False    False                    False   
1       False     True    False    False                    False   
2       False    False    False    False                    False   
3       False    False    False    False                    False   
4       False     True    False    False                    False   
...       ...      ...      ...      ...                      ...   
8688    False    False    False    False                     True   
8689    False    False     True    False                    False   
8690    False    False     True    False                    False   
8691     True    False    False    False                     True   
8692     True    False    False    False                    False   

      Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                         False                     True       True     False  
1                         False                     True       True     False  
2                         False                     True      False      True  
3                         False                     True       True     False  
4                         False                     True       True     False  
...                         ...                      ...        ...       ...  
8688                      False                    False      False      True  
8689                       True                    False       True     False  
8690                      False                     True       True     False  
8691                      False                    False       True     False  
8692                      False                     True       True     False  

[6606 rows x 25 columns]

In [12]:
#your code here
spaceship_numerical = spaceship_cleaned.select_dtypes(include=['float64','bool'])

# Imprimir el DataFrame solo con columnas numéricas
print(spaceship_numerical)

       Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0     39.0          0.0        0.0           0.0     0.0     0.0        False   
1     24.0        109.0        9.0          25.0   549.0    44.0         True   
2     58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3     33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4     16.0        303.0       70.0         151.0   565.0     2.0         True   
...    ...          ...        ...           ...     ...     ...          ...   
8688  41.0          0.0     6819.0           0.0  1643.0    74.0        False   
8689  18.0          0.0        0.0           0.0     0.0     0.0        False   
8690  26.0          0.0        0.0        1872.0     1.0     0.0         True   
8691  32.0          0.0     1049.0           0.0   353.0  3235.0        False   
8692  44.0        126.0     4688.0           0.0     0.0    12.0         True   

      HomePlanet_Earth  Hom

**Perform Train Test Split**

In [13]:
features = numerical_spaceship.drop(columns = ["Transported"])
target = numerical_spaceship["Transported"]

In [14]:
#your code here
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.20, random_state=0)

**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

In [15]:
# normalizar

In [16]:
normalizer = MinMaxScaler()

normalizer.fit(X_train)

MinMaxScaler()

In [17]:
X_train_norm = normalizer.transform(X_train)

X_test_norm = normalizer.transform(X_test)

## Bagging and Pasting

In [18]:
bagging_reg = BaggingRegressor(DecisionTreeRegressor(max_depth=20),
                               n_estimators=100,
                               max_samples = 1000)

In [19]:
bagging_reg.fit(X_train_norm, y_train)

BaggingRegressor(estimator=DecisionTreeRegressor(max_depth=20),
                 max_samples=1000, n_estimators=100)

Evaluate model's performance

In [20]:
pred = bagging_reg.predict(X_test_norm)

print("Bagging and Pasting MAE", mean_absolute_error(pred, y_test))
print("Bagging and Pasting RMSE", mean_squared_error(pred, y_test, squared=False))
print("Bagging and Pasting R2 score", bagging_reg.score(X_test_norm, y_test))

Bagging and Pasting MAE 0.2778386375560917
Bagging and Pasting RMSE 0.378354598982659
Bagging and Pasting R2 score 0.42739118971468515


## Random Forests

In [21]:
forest = RandomForestRegressor(n_estimators=100,
                             max_depth=20)

- Training the model

In [22]:
forest.fit(X_train_norm, y_train)

RandomForestRegressor(max_depth=20)

- Evaluate model's performance

In [23]:
pred = forest.predict(X_test_norm)

print("Random Forests MAE", mean_absolute_error(pred, y_test))
print("Random Forests RMSE", mean_squared_error(pred, y_test, squared=False))
print("Random Forests R2 score", forest.score(X_test_norm, y_test))

Random Forests MAE 0.26843460982163697
Random Forests RMSE 0.38222578581019206
Random Forests R2 score 0.4156137946471248


## Gradient Boosting

- Initialize a AdaBoost model

In [24]:
gb_reg = GradientBoostingRegressor(max_depth=20,
                                   n_estimators=100)

- Training the model

In [25]:
gb_reg.fit(X_train_norm, y_train)

GradientBoostingRegressor(max_depth=20)

Evaluate model's performance

In [26]:
pred = gb_reg.predict(X_test_norm)

print("Gradient Boosting MAE", mean_absolute_error(pred, y_test))
print("Gradient Boosting RMSE", mean_squared_error(pred, y_test, squared=False))
print("Gradient Boosting R2 score", gb_reg.score(X_test_norm, y_test))

Gradient Boosting MAE 0.2627483919569256
Gradient Boosting RMSE 0.41906419652472743
Gradient Boosting R2 score 0.2975407967643384


## Adaptive Boosting

- Initialize a AdaBoost model

In [27]:
ada_reg = AdaBoostRegressor(DecisionTreeRegressor(max_depth=20),
                            n_estimators=100)

- Training the model

In [28]:
ada_reg.fit(X_train_norm, y_train)

AdaBoostRegressor(estimator=DecisionTreeRegressor(max_depth=20),
                  n_estimators=100)

Evaluate model's performance

In [29]:
pred = ada_reg.predict(X_test_norm)

print("Adaptive Boosting MAE", mean_absolute_error(pred, y_test))
print("Adaptive Boosting RMSE", mean_squared_error(pred, y_test, squared=False))
print("Adaptive Boosting R2 score", ada_reg.score(X_test_norm, y_test))

Adaptive Boosting MAE 0.24566626933279767
Adaptive Boosting RMSE 0.4295454255815725
Adaptive Boosting R2 score 0.261962909447783


Which model is the best and why?

Bagging and Pasting MAE 0.2778386375560917
Bagging and Pasting RMSE 0.378354598982659
Bagging and Pasting R2 score 0.42739118971468515

EL mejor modelo los anteriores en mi caso es Bagging and Pasting por que son aquellos los que resultados de las metricas son mejores los resultados 